[wasm] dotnet.js modularization

pavelsavara opened the pull request:
### Context
After [dotnet.js was modularized](https://github.com/dotnet/runtime/pull/61313), Blazor could start using the new API we expose. 

This PR updates Blazor to use new API [createDotnetRuntime()](https://github.com/dotnet/runtime/blob/main/src/mono/wasm/runtime/dotnet.d.ts)

The current default output of runtime is CommonJS module, which is at the same time backward compatible with .NET 6.0 `globalThis.Module` style. In the future ES6 would be default module format. This change also makes Blazor future-compatible with ES6 runtime module.

### Changes
- use dynamic `import` instead of HTML `<script>` tag to load dotnet
- updated `createEmscriptenModuleInstance` to use the new `createDotnetRuntime()` API
- made import of dotnet.js parallel with loading other assets.

- added `src/Components/Web.JS/@types/dotnet` temporarily, until we figure out how to pull this out of `Microsoft.NETCore.App.Runtime.Mono.browser-wasm.7.0.0-dev.nupkg`
- removed @types/emscripten type definitions
- fixed usages of `addRunDependency`, `removeRunDependency`
- mapped `System_Object` and other types to `MonoObject` etc
- global `Module` is no longer needed by runtime
    - therefore removed `__wasmmodulecallback__`
    - Unfortunately I have no way of testing the platform which had this problem in the past.
- because it's not ES6 yet
    - there is helper callback `__onDotnetRuntimeLoaded` to get the API when the CJS module is loaded
- removed few `cwrap` calls and replaced them with API usage
- Implemented DRI for dynamic import via link `modulepreload`


### Feedback

This is also opportunity for Blazor team to give feedback about the new API.
There are more samples of the API usage here https://github.com/dotnet/runtime/pull/62292

Fixes https://github.com/dotnet/aspnetcore/issues/39283
Fixes: https://github.com/dotnet/aspnetcore/issues/36805

pavelsavara commented:
cc @lewing

dougbu commented on global.json:
```diff
@@ -1,9 +1,9 @@
 {
   "sdk": {
-    "version": "7.0.100-alpha.1.21613.8"
+    "version": "7.0.100-alpha.1.21617.5"
```
changes here will conflict w/ #39127 
Suggest waiting for that to get in and handling the conflict

pavelsavara replied:
Sure, thanks!

pavelsavara committed:
- use import to load dotnet.js instead of html tag
- updated createEmscriptenModuleInstance to use the new createDotnetRuntime() API

- removed @types/emscripten type definitions
- added src/Components/Web.JS/@types/dotnet temporarily, until we figure out how to pull this out of Microsoft.NETCore.App.Runtime.Mono.browser-wasm.7.0.0-dev.nupkg
- fixed usages of addRunDependency, removeRunDependency
- mapped System_Object and other types to MonoObject etc
- global Module is no longer needed by runtime
- therefore removed __wasmmodulecallback__
- removed few cwrap calls and replaced them with API usage

wtgodbe reviewed (approved):
LGTM from a build perspective, but it'd be good to get approval from someone on Blazor too

TanayParikh commented on src/Components/Web.JS/@types/dotnet/package.json:
```diff
@@ -0,0 +1,34 @@
+{
+  "name": "dotnet-runtime",
+  "description": ".NET is a developer platform with tools and libraries for building any type of app, including web, mobile, desktop, games, IoT, cloud, and microservices.",
```
```suggestion
  "name": "dotnet-runtime",
  "private": true,
  "description": ".NET is a developer platform with tools and libraries for building any type of app, including web, mobile, desktop, games, IoT, cloud, and microservices.",
```

TanayParikh commented on src/Components/Web.JS/src/Platform/Mono/MonoPlatform.ts:
```diff
@@ -181,66 +174,59 @@ function addScriptTagsToDocument(resourceLoader: WebAssemblyResourceLoader) {
-  }
+  let src = `./${dotnetJsResourceName}`;
+
+  // TODO How to do SRI dotnetJsContentHash with dynamic import ?
```
Is this something we need to address before merging?

pavelsavara replied:
Not from my perspective, but I don't know what impact it has downstream.

TanayParikh replied:
Discussed offline with @pavelsavara, we don't believe this'll be necessary.

```suggestion
```

pavelsavara committed:
Merge branch 'main' into wasm_modularization_1

# Conflicts:
#	src/Components/Web.JS/dist/Release/blazor.server.js
#	src/Components/Web.JS/dist/Release/blazor.webview.js
#	src/Components/Web.JS/src/Platform/Mono/MonoTypes.ts
#	src/Components/Web.JS/src/Platform/Platform.ts

TanayParikh commented on src/Components/Web.JS/src/Platform/Mono/MonoPlatform.ts:
```diff
@@ -181,66 +174,59 @@ function addScriptTagsToDocument(resourceLoader: WebAssemblyResourceLoader) {
-  window['__wasmmodulecallback__'] = callback;
+async function createEmscriptenModuleInstance(resourceLoader: WebAssemblyResourceLoader, onReady: () => void, onError: (reason?: any) => void) {
 
-  // Note: Any updates to the following script will require updating the inline script hash if using CSP
-  scriptElem.text = 'var Module; window.__wasmmodulecallback__(); delete window.__wasmmodulecallback__;';
```
🎉  this removal is great! Fixes: https://github.com/dotnet/aspnetcore/issues/36805

Did you get a chance to validate it using MacOS Catalina?

pavelsavara replied:
Nope, I don't have access to that device.

TanayParikh replied:
Ah okay, let me give it a try...

TanayParikh replied:
Just tried it on my Mac (running Big Sur -> successor to Catalina). Didn't encounter any issues. I don't have a Catalina device to verify on. 

Given `This may be fixed in a later version of macOS/iOS`, I'm going to consider this resolved now.

TanayParikh commented on src/Components/Web.JS/src/Platform/WebAssemblyConfigLoader.ts:
```diff
@@ -1,6 +1,7 @@
 import { BootConfigResult } from './BootConfig';
 import { System_String, System_Object } from './Platform';
 import { Blazor } from '../GlobalExports';
+import { BINDING } from './Mono/MonoPlatform';
```
Is this needed?

```suggestion
```

pavelsavara replied:
Yes, it's used 3 times in the file.

TanayParikh reviewed (commented):
Looking good, some minor nits. 

Tried resolving the merge conflicts, but I don't have perms to push onto your fork.

pavelsavara committed:
fix merge

pavelsavara committed:
private

Co-authored-by: Tanay Parikh <TanayParikh@users.noreply.github.com>

TanayParikh reviewed (approved):
Thanks @pavelsavara this looks good to me. @pranavkm any thoughts or concerns before merging?

pranavkm commented:
We should let @SteveSandersonMS have a look at this since he might have specific opinions about it. I've assigned the PR to him to merge.

pavelsavara committed:
Update src/Components/Web.JS/src/Platform/Mono/MonoPlatform.ts

Co-authored-by: Tanay Parikh <TanayParikh@users.noreply.github.com>

pavelsavara committed:
Merge branch 'main' into wasm_modularization_1

pavelsavara committed:
update dotnet.d.ts

pavelsavara committed:
fix build

pavelsavara commented on src/Components/Web.JS/src/Platform/Mono/MonoPlatform.ts:
```diff
@@ -49,19 +52,9 @@ export const monoPlatform: Platform = {
-        window['Module'] = createEmscriptenModuleInstance(resourceLoader, resolve, reject);
-        addScriptTagsToDocument(resourceLoader);
-      });
+      createEmscriptenModuleInstance(resourceLoader, resolve, reject);
```
This is regular promse now.

javiercn replied:
Do we know for sure this works well with Mac OS? (Note that our CI doesn't have coverage for this scenario). I believe the comment no longer applies, but want to make sure that's the case.

pavelsavara commented on src/Components/Web.JS/src/Platform/Mono/MonoPlatform.ts:
```diff
@@ -183,66 +176,58 @@ function addScriptTagsToDocument(resourceLoader: WebAssemblyResourceLoader) {
+  const dotnetJsBeingLoaded = importDotnetJs(resourceLoader);
   const resources = resourceLoader.bootConfig.resources;
-  const module = (window['Module'] || {}) as typeof Module;
+  const moduleConfig = (window['Module'] || {}) as typeof Module;
```
This is for your consideration if your users expect and use `window.Module`. From runtime perspective, you could stop poluting `window`.

javiercn replied:
This is something that ideally we want to cleanup by default, but I do believe we need an option for backwards compatibility. We can flow this via blazorwebassemblyoptions and have it be opt-in (to ease out anyone using this from it)

SteveSandersonMS replied:
+1 on dropping the global by default and having an opt-in to bring it back.

My only concern would be about discoverability. If people upgrade an app and it starts failing with `Uncaught ReferenceError: Module is not defined`, it doesn't give them any clue that there's a switch to fix it. Not sure how to solve that.

pavelsavara replied:
I'm interested to listen on that design conversation, when you guys have it. It's tough problem.

We solved similar problem by polluting global namespace with `getDotnetRuntime(runtimeId: number)`. Our debugger now attaches to id `0`, assuming that there is at least one runtime on the page and that's the one which should be debugged.

This function could be also used to locate existing runtime and use it, instead of creating another instance on the page. If you have multiple WebComponents which don't know about each other, for example. Runtime has version exposed, so that you could know if it's compatible with you. But it doesn't tell you what AOT stripped.

pavelsavara commented on src/Components/Web.JS/src/Platform/Mono/MonoPlatform.ts:
```diff
@@ -278,220 +263,229 @@ function createEmscriptenModuleInstance(resourceLoader: WebAssemblyResourceLoade
-        throw ex;
+  const createDotnetRuntime = await dotnetJsBeingLoaded;
+
+  return await createDotnetRuntime(({ MONO: mono, BINDING: binding, Module: module }) => {
```
This returns promise, resolved when the runtime started.

pavelsavara commented on src/Components/Web.JS/src/Platform/Mono/MonoPlatform.ts:
```diff
@@ -278,220 +263,229 @@ function createEmscriptenModuleInstance(resourceLoader: WebAssemblyResourceLoade
-    }
+    const dotnetModuleConfig: DotnetModuleConfig = {
+      ...moduleConfig,
+      disableDotnet6Compatibility: false,
```
We are still exporting methods and objects to `window` or `globalThis` when this is `false`. 
That is bad practice, but backward compatible.
You could consider if your users expect those symbols to be there in Net7.0 or if you are OK breaking those undocumented assumptions.

SteveSandersonMS replied:
Agreed on the goals. We have more work to do if we want Blazor to be loadable side-by-side, as (eg) the JS interop APIs can be called from global scope without saying which .NET instance you're trying to call. We'd need to make breaking APIs changes if we want side-by-side by default.

pavelsavara commented on src/Components/Web.JS/src/Platform/Mono/MonoPlatform.ts:
```diff
@@ -183,66 +176,58 @@ function addScriptTagsToDocument(resourceLoader: WebAssemblyResourceLoader) {
-
-  // For consistency with WebAssemblyResourceLoader, we only enforce SRI if caching is allowed
-  if (resourceLoader.bootConfig.cacheBootResources) {
-    scriptElem.integrity = dotnetJsContentHash;
```
There is no way how to check SRI via dynamic `import` AFAIK.
There are already other dynamic imports in the codebase.

javiercn replied:
I think we might want to produce a single bundle that incorporates dotnet.js into blazor.webassembly.js we don't want to loose the SRI bit (even though it's not that important for scripts hosted locally) and consuming the module and producing a single bundle is more natural. (That's the direction that we were headed).

That said, I think the current changes are a good first step, and we need to think a few things through to achieve the rest:
* Single bundle (we need to think about how this plays with relinking of dotnet.js)
* Global state: I think we need to figure out the interface between blazor/mono when we use modules. Removing any global state is something we want to achieve, that said, I think we need to make it configurable for backwards compatibility. 
  * I think this applies to both Blazor and mono globals
* Contract between Blazor/mono: I believe we were doing some things through global callbacks, it would be neat if we are able to get rid of those.

pavelsavara replied:
- runtime doesn't have any global state anymore, it's fully modularized. I should probably write unit test which creates multiple runtimes on same page to prove it.
- if you set `disableDotnet6Compatibility: true` you will see if any of your tests fail, because blazor still calls runtime via global namespace.

pavelsavara replied:
Note that packing dotnet into `blazor.webassembly.js` is possible, but it has 2 consequences
- it tries to find `dotnet.wasm` in same directory as the bundle, which is not the same as location of eventual npm package. In case of Blazor this is probably non-issue as you have your cache anyway.
- AOT/native linking scenarios related to `dotnet.wasm` and `dotnet.js` need to be designed.

Anyway, trivial [webpack sample here](https://github.com/dotnet/runtime/tree/3591addba0a7394123fd522c9951d52c5854f261/src/mono/sample/wasm/browser-webpack)

pavelsavara replied:
Let's create separate issue and merge this one. I would like to proceed with ES6 switch and let you do Blazor cleanup.

pavelsavara replied:
@SteveSandersonMS do you think that the `modulepreload` would prevent `import` from loading file with different hash ?

SteveSandersonMS replied:
The sample at https://stackoverflow.com/a/63029759 appears to show that it does prevent loading a file with a wrong hash.

pavelsavara replied:
I tested it and it works. Thanks! 
Code is updated and I do it only for chromium.

javiercn commented:
@pavelsavara thanks for taking on this. We've been thinking about switching to this approach on our end too for various reasons. I will be adding some comments as I review things.

javiercn reviewed (commented):
Changes look great!

I have a few questions, but nothing critical, this seems like a step on the right direction!

SteveSandersonMS commented on src/Components/Web.JS/src/Platform/Mono/MonoPlatform.ts:
```diff
@@ -10,10 +10,13 @@ import { Platform, System_Array, Pointer, System_Object, System_String, HeapLock
 
-declare let Module: EmscriptenModule;
+// initially undefined and only fully initialized after createEmscriptenModuleInstance()
+export let BINDING: BINDINGType = <any>undefined;
```
I appreciate the value in not consuming them as globals here.

Just to check though: by default, they will still exist as globals, right? Third-party code will take a dependency on this today.

pavelsavara replied:
Yes, as far as you set `disableDotnet6Compatibility: false`

SteveSandersonMS commented on src/Components/Web.JS/src/Platform/Mono/MonoPlatform.ts:
```diff
@@ -10,10 +10,13 @@ import { Platform, System_Array, Pointer, System_Object, System_String, HeapLock
-declare let Module: EmscriptenModule;
+// initially undefined and only fully initialized after createEmscriptenModuleInstance()
+export let BINDING: BINDINGType = <any>undefined;
+export let MONO: MONOType = <any>undefined;
```
Shouldn't it be of type `MONOType | undefined` if it's sometimes undefined? Would that help other code to keep track of checking that it's become defined before using it?

Same with `BINDING` above.

SteveSandersonMS replied:
Actually I see how it would cause a lot of noise in the code below if you did that. The way you've done it looks good.

pavelsavara replied:
Would you like to make asserts in each method ? If so, it's easy to do.

SteveSandersonMS replied:
I don't think we want to introduce any extra runtime checking in such low-level code (e.g., reading bytes from memory). So I'd vote for leaving this aspect as-is.

SteveSandersonMS commented on src/Components/Web.JS/src/Platform/Mono/MonoPlatform.ts:
```diff
@@ -24,13 +27,13 @@ let currentHeapLock: MonoHeapLock | null = null;
 // except without having to parse the 'type' parameter, and with less risk of mistakes at the call site
 function getValueI16(ptr: number) {
-  return Module.HEAP16[ptr >> 1];
+  return MONO.getI16(ptr);
```
The wrappers here, `getValueI16`, `getValueI32`, etc., no longer seem to do anything - would it be viable to remove them and inline them into their usage sites? We need to keep `readInt16Field` etc, but it could directly call `MONO.getI16`.

SteveSandersonMS commented on src/Components/Web.JS/src/Platform/Mono/MonoPlatform.ts:
```diff
@@ -49,19 +52,9 @@ export const monoPlatform: Platform = {
+      createEmscriptenModuleInstance(resourceLoader, resolve, reject);
+    }).then(() => {
+      window['Module'] = Module;
     });
```
Instead of using `then` explicitly, could we use `await`? Just marginally cleaner.

For example, instead of returning `new Promise`, we could do something like:

```ts
return (async () => {
   ... code involving await ...
})();
```

Also I don't quite understand the original code because it's not returning the promise from `createEmscriptenModuleInstance`, so it will just call the `then` callback immediately won't it? Is that a typo?

pavelsavara replied:
It's the `onReady();` executed from emscripten callback, see line 450. We could move the `new Promise` inside createEmscriptenModuleInstance, but it's not changing much about it.

SteveSandersonMS replied:
That clarifies things, thanks. I do think it would be much cleaner to make `createEmscriptenModuleInstance` return a regular promise (just by being an async function) instead of manually passing and calling `resolve`/`reject`.

pavelsavara replied:
Improved

SteveSandersonMS commented on src/Components/Web.JS/src/Platform/Mono/MonoPlatform.ts:
```diff
@@ -183,66 +176,58 @@ function addScriptTagsToDocument(resourceLoader: WebAssemblyResourceLoader) {
-    scriptElem.crossOrigin = 'anonymous';
-  }
+  let src = `./${dotnetJsResourceName}`;
+
```
```suggestion
```

SteveSandersonMS commented on src/Components/Web.JS/src/Platform/Mono/MonoPlatform.ts:
```diff
@@ -183,66 +176,58 @@ function addScriptTagsToDocument(resourceLoader: WebAssemblyResourceLoader) {
+    const customSrc = resourceLoader.startOptions.loadBootResource(resourceType, dotnetJsResourceName, src, dotnetJsContentHash);
     if (typeof (customSrc) === 'string') {
-      scriptElem.src = customSrc;
+      src = customSrc;
```
Are you sure this works? It looks like it would be a breaking change. If an existing app returns a URL that doesn't start with `./`, then it won't work with `import`.

pavelsavara replied:
Changed it to `new URL(`_framework/${dotnetJsResourceName}`, document.baseURI)` hope that's better

SteveSandersonMS commented on src/Components/Web.JS/src/Platform/Mono/MonoPlatform.ts:
```diff
@@ -183,66 +176,58 @@ function addScriptTagsToDocument(resourceLoader: WebAssemblyResourceLoader) {
-      scriptElem.src = customSrc;
+      src = customSrc;
     } else if (customSrc) {
       // Since we must load this via a <script> tag, it's only valid to supply a URI (and not a Request, say)
```
This comment is outdated now

pavelsavara replied:
will fix

SteveSandersonMS commented on src/Components/Web.JS/src/Platform/Mono/MonoPlatform.ts:
```diff
@@ -183,66 +176,58 @@ function addScriptTagsToDocument(resourceLoader: WebAssemblyResourceLoader) {
+      resolve(createDotnetRuntime);
+    };
+    // this tells webpack to do dynamic import at runtime, not at compile time
+    import(/* webpackIgnore: true */ src).then(({ default: createDotnetRuntime }) => {
```
As with the other suggestion above, would it be possible to make this use `await` instead of `then`? It just make the code easier for us to maintain in the future, e.g., if someone comes to add in a `try`/`catch` around it.

pavelsavara replied:
will do

pavelsavara committed:
Update src/Components/Web.JS/src/Platform/Mono/MonoPlatform.ts

Co-authored-by: Steve Sanderson <SteveSandersonMS@users.noreply.github.com>

SteveSandersonMS commented on src/Components/Web.JS/src/Platform/Mono/MonoPlatform.ts:
```diff
@@ -49,19 +52,9 @@ export const monoPlatform: Platform = {
-      });
+      createEmscriptenModuleInstance(resourceLoader, resolve, reject);
+    }).then(() => {
+      window['Module'] = Module;
```
Is this necessary? Doesn't `disableDotnet6Compatibility: false` do it anyway?

pavelsavara replied:
You are right

SteveSandersonMS commented on src/Components/Web.JS/src/Platform/Mono/MonoPlatform.ts:
```diff
@@ -278,220 +263,229 @@ function createEmscriptenModuleInstance(resourceLoader: WebAssemblyResourceLoade
+        // Load it into the Mono runtime
+        const data = new Uint8Array(dataBuffer);
+        const heapAddress = Module._malloc(data.length);
+        const heapMemory = new Uint8Array(Module.HEAPU8.buffer, <any>heapAddress, data.length);
```
```suggestion
        const heapMemory = new Uint8Array(Module.HEAPU8.buffer, heapAddress as any, data.length);
```

No big deal - it's just the style we use elsewhere.

pavelsavara replied:
malloc now returns type `VoidPtr` which `new Uint8Array` doesn't know about.

pavelsavara replied:
Will do.

SteveSandersonMS commented:
Thanks for this, @pavelsavara! This looks really great. There are a few minor details to resolve (as per comments) but the overall change is very much appreciated and valuable.

pavelsavara committed:
feedback

SteveSandersonMS commented on src/Components/Web.JS/src/Platform/Mono/MonoPlatform.ts:
```diff
@@ -176,39 +174,49 @@ function importDotnetJs(resourceLoader: WebAssemblyResourceLoader): Promise<Crea
     .filter(n => n.startsWith('dotnet.') && n.endsWith('.js'))[0];
   const dotnetJsContentHash = resourceLoader.bootConfig.resources.runtime[dotnetJsResourceName];
-  let src = `./${dotnetJsResourceName}`;
+  let src = new URL(`_framework/${dotnetJsResourceName}`, document.baseURI);
```
Sorry, this is another breaking change. The `loadBootResource` callback needs to receive a string, otherwise existing code that does things like `value.endsWith(...)` will start failing.

I think we should continue to pass `_framework/blazor.webassembly.js` (not `./_framework/blazor.webassembly.js`) into `loadBootResource`, and then normalize the result against baseURI afterwards.

pavelsavara replied:
I'm confused now, this is about `dotnet.*.js` not about `blazor.webassembly.js`. Also this new version doesn't pass `./` anymore. Also this is not about `resourceLoader`, we are not caching it at all, only browser does.
Could you please elaborate or propose a fix ?

pavelsavara committed:
implement SRI via link modulepreload

pavelsavara committed:
fix

pavelsavara committed:
fix

SteveSandersonMS commented on src/Components/Web.JS/src/Platform/Mono/MonoPlatform.ts:
```diff
@@ -183,66 +174,88 @@ function addScriptTagsToDocument(resourceLoader: WebAssemblyResourceLoader) {
+    try {
+      if (globalThis.navigator) {
+        const nav: any = globalThis.navigator;
+        if (nav.userAgentData && nav.userAgentData.brands && nav.userAgentData.brands.some((i: any) => i.brand === 'Chromium')) {
```
Could you clarify if it's strictly necessary to limit this to Chromium? I know the enforcement only takes effect on Chromium, but is there any harm in adding the link tag regardless on other browsers? Other browsers may add support for this in the future, and it would be ideal if we didn't need to maintain knowledge about which code paths ran on which browsers.

Because we're so early in the 7.0 cycle, we can afford some level of risk here, so we don't need to limit this just to Chromium. It's likely that adding the `<link rel=preload>` will have no adverse effect on other browsers, and if it does, we have plenty of time to figure that out and adapt.

In summary, if we can just remove browser checking and nothing blows up on Firefox/Safari, then that sounds better to me.

pavelsavara replied:
Done, I tested with FF. It doesn't validate it, but it doesn't break.

SteveSandersonMS commented on src/Components/Web.JS/src/Platform/Mono/MonoPlatform.ts:
```diff
@@ -183,66 +174,88 @@ function addScriptTagsToDocument(resourceLoader: WebAssemblyResourceLoader) {
+        }
+      }
+    } catch (err) {
+      console.error(err);
```
I expect this was left over from development. We should remove the `try`/`catch` and the logging to console, because if there's an unhandled error we'd prefer the promise chain to be rejected and have the natural unhandled error behavior.

pavelsavara replied:
will do

SteveSandersonMS commented on src/Components/Web.JS/src/Platform/Mono/MonoPlatform.ts:
```diff
@@ -183,66 +174,88 @@ function addScriptTagsToDocument(resourceLoader: WebAssemblyResourceLoader) {
+          scriptElem.crossOrigin = 'anonymous';
+          document.head.appendChild(scriptElem);
+        }
+      }
```
Shouldn't there be an `else` case for non-Chrome browsers? This requirement goes away if we just run the same code on all browsers of course.

SteveSandersonMS replied:
No sorry I'm wrong here. Deleting!

SteveSandersonMS commented on src/Components/Web.JS/src/Platform/Mono/MonoPlatform.ts:
```diff
@@ -183,66 +174,88 @@ function addScriptTagsToDocument(resourceLoader: WebAssemblyResourceLoader) {
-  window['__wasmmodulecallback__'] = callback;
+  // GOTCHA: remove this once runtime switched to ES6
+  // this is capturing the export via callback we have in CJS version of the runtime
+  let cjsExportResolve: (data: CreateDotnetRuntimeType) => void = <any>undefined;
```
```suggestion
  let cjsExportResolve: (data: CreateDotnetRuntimeType) => void = undefined as any;
```

SteveSandersonMS commented on src/Components/Web.JS/src/Platform/Mono/MonoPlatform.ts:
```diff
@@ -278,220 +291,232 @@ function createEmscriptenModuleInstance(resourceLoader: WebAssemblyResourceLoade
+                BINDING.mono_obj_array_set(pdbBytes, i, BINDING.js_typed_array_to_array(pdb));
               }
-              return array;
+              return <any>pdbBytes;
```
```suggestion
              return pdbBytes as any;
```

SteveSandersonMS commented on src/Components/Web.JS/src/Platform/Mono/MonoPlatform.ts:
```diff
@@ -278,220 +291,232 @@ function createEmscriptenModuleInstance(resourceLoader: WebAssemblyResourceLoade
+                const assembly = assemblies[i] as ArrayBuffer;
+                BINDING.mono_obj_array_set(assemblyBytes, i, BINDING.js_typed_array_to_array(new Uint8Array(assembly)));
+              }
+              return <any>assemblyBytes;
```
```suggestion
              return assemblyBytes as any;
```

SteveSandersonMS commented on src/Components/Web.JS/src/Platform/Mono/MonoPlatform.ts:
```diff
@@ -183,66 +174,88 @@ function addScriptTagsToDocument(resourceLoader: WebAssemblyResourceLoader) {
+    runtimeReadyResolve = resolve;
+    runtimeReadyReject = reject;
+  });
+
```
```suggestion
```

SteveSandersonMS commented on src/Components/Web.JS/src/Platform/Mono/MonoPlatform.ts:
```diff
@@ -10,10 +12,13 @@ import { Platform, System_Array, Pointer, System_Object, System_String, HeapLock
 
-declare let Module: EmscriptenModule;
+// initially undefined and only fully initialized after createEmscriptenModuleInstance()
+export let BINDING: BINDINGType = <any>undefined;
```
```suggestion
export let BINDING: BINDINGType = undefined as any;
```

SteveSandersonMS commented on src/Components/Web.JS/src/Platform/Mono/MonoPlatform.ts:
```diff
@@ -10,10 +12,13 @@ import { Platform, System_Array, Pointer, System_Object, System_String, HeapLock
-declare let Module: EmscriptenModule;
+// initially undefined and only fully initialized after createEmscriptenModuleInstance()
+export let BINDING: BINDINGType = <any>undefined;
+export let MONO: MONOType = <any>undefined;
```
```suggestion
export let MONO: MONOType = undefined as any;
```

SteveSandersonMS commented on src/Components/Web.JS/src/Platform/Mono/MonoPlatform.ts:
```diff
@@ -10,10 +12,13 @@ import { Platform, System_Array, Pointer, System_Object, System_String, HeapLock
+// initially undefined and only fully initialized after createEmscriptenModuleInstance()
+export let BINDING: BINDINGType = <any>undefined;
+export let MONO: MONOType = <any>undefined;
+export let Module: DotnetModuleConfig & EmscriptenModule = <any>undefined;
```
```suggestion
export let Module: DotnetModuleConfig & EmscriptenModule = undefined as any;
```

SteveSandersonMS commented:
Thanks @pavelsavara - I think we're just about there!

I added a couple of substantive comments about the new link modulepreload code, plus a bunch of trivial syntax tweak suggestions. Hope that's OK. Once that's resolved I think this can be merged!

pavelsavara committed:
Update src/Components/Web.JS/src/Platform/Mono/MonoPlatform.ts

Co-authored-by: Steve Sanderson <SteveSandersonMS@users.noreply.github.com>

pavelsavara committed:
Update src/Components/Web.JS/src/Platform/Mono/MonoPlatform.ts

Co-authored-by: Steve Sanderson <SteveSandersonMS@users.noreply.github.com>

pavelsavara committed:
Update src/Components/Web.JS/src/Platform/Mono/MonoPlatform.ts

Co-authored-by: Steve Sanderson <SteveSandersonMS@users.noreply.github.com>

pavelsavara committed:
Update src/Components/Web.JS/src/Platform/Mono/MonoPlatform.ts

Co-authored-by: Steve Sanderson <SteveSandersonMS@users.noreply.github.com>

pavelsavara committed:
Update src/Components/Web.JS/src/Platform/Mono/MonoPlatform.ts

Co-authored-by: Steve Sanderson <SteveSandersonMS@users.noreply.github.com>

pavelsavara committed:
Update src/Components/Web.JS/src/Platform/Mono/MonoPlatform.ts

Co-authored-by: Steve Sanderson <SteveSandersonMS@users.noreply.github.com>

pavelsavara committed:
Update src/Components/Web.JS/src/Platform/Mono/MonoPlatform.ts

Co-authored-by: Steve Sanderson <SteveSandersonMS@users.noreply.github.com>

pavelsavara committed:
feedback

SteveSandersonMS reviewed (approved):
Awesome, thanks so much for the updates!

pavelsavara commented:
There is some issue with IIS test, I think it's unrelated and it could be merged.
